cdap-app-fabric/src/main/java/io/cdap/cdap/app/deploy/AppDeletionSubscriberService.java

```diff
@@ -0,0 +1,132 @@
+
+/**
+ * A TMS Subscriber Service responsible for consuming app deletion messages from TMS and executes removal from
+ * the app spec table. Subscribes to the Constants.AppFabric.APP_DELETION_EVENT_TOPIC topic.
+ */
+public class AppDeletionSubscriberService extends AbstractEventSubscriberService<AppDeletionMessage> {
+  private static final String SUBSCRIBER = "appdelete.";
+  private static final Gson GSON = new GsonBuilder()
+    .registerTypeAdapter(EntityId.class, new EntityIdTypeAdapter())
+    .create();
+
+  private final CConfiguration cConf;
+  private final MultiThreadMessagingContext messagingContext;
+  private final TransactionRunner transactionRunner;
+  private final int maxRetriesOnConflict;
+  private final MetricsCollectionService metricsCollectionService;
+
+
+  @Inject
+  AppDeletionSubscriberService(CConfiguration cConf, MessagingService messagingService,
+                            MetricsCollectionService metricsCollectionService,
+                            TransactionRunner transactionRunner) {
+    super(
+      NamespaceId.SYSTEM.topic(cConf.get(Constants.AppFabric.APP_DELETION_EVENT_TOPIC)),
+      cConf.getInt(Constants.AppFabric.APP_DELETE_EVENT_FETCH_SIZE),
+      cConf.getInt(TxConstants.Manager.CFG_TX_TIMEOUT),
+      cConf.getLong(Constants.AppFabric.APP_DELETE_EVENT_POLL_DELAY_MILLIS),
+      RetryStrategies.fromConfiguration(cConf, "app.delete.event."),
+      metricsCollectionService.getContext(ImmutableMap.of(
+        Constants.Metrics.Tag.COMPONENT, Constants.Service.MASTER_SERVICES,
+        Constants.Metrics.Tag.INSTANCE_ID, "0",
+        Constants.Metrics.Tag.NAMESPACE, NamespaceId.SYSTEM.getNamespace(),
+        Constants.Metrics.Tag.TOPIC, cConf.get(Constants.AppFabric.APP_DELETION_EVENT_TOPIC),
+        Constants.Metrics.Tag.CONSUMER, SUBSCRIBER
+      )));
+
+    this.cConf = cConf;
+    this.messagingContext = new MultiThreadMessagingContext(messagingService);
+    this.transactionRunner = transactionRunner;
+    this.maxRetriesOnConflict = cConf.getInt(Constants.Metadata.MESSAGING_RETRIES_ON_CONFLICT);
+    this.metricsCollectionService = metricsCollectionService;
+  }
+
+  @Override
+  protected TransactionRunner getTransactionRunner() {
+    return transactionRunner;
+  }
+
+  @Override
+  protected void processMessages(Iterator<ImmutablePair<String, AppDeletionMessage>> messages) throws Exception {
+    while (messages.hasNext()) {
+      ImmutablePair<String, AppDeletionMessage> pair = messages.next();
+      AppDeletionMessage message = pair.getSecond();
+      ApplicationId appId = message.getApplicationId();
+      removeApp(appId);
+    }
+  }
+
+  private void removeApp(ApplicationId id) throws IOException {
+    TransactionRunners.run(transactionRunner, context -> {
+      getAppStateTable(context).deleteAll(id.getNamespaceId(), id.getApplication());
+      AppMetadataStore metaStore = getAppMetadataStore(context);
+      metaStore.deleteApplication(id.getAppReference());
```

Comment:
can we combine all of these into a single method in AppMetadataStore? The subscriber shouldn't really have special knowledge about what is everything version related in the store. 

Otherwise it will be easy to add something to the store in the future and forget to delete it here.

Reply:
@albertshau do we want to move markDeleteApplication here https://github.com/cdapio/cdap/blob/develop/cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/store/DefaultStore.java#L632 (removeApplication) replacing the current logic, since the subscriber directly calls AppMetadataStore ?

Reply:
yeah, I think you can have markApplicationDeleted to the AppMetadataStore (that DefaultStore will use) and have a removeApplicationData() or however you want to call it that will be called here.